---
title: 'Filter and prepare analysis ready variants'
teaching: 10
exercises: 2
---

:::::::::::::::::::::::::::::::::::::: questions 

- How

::::::::::::::::::::::::::::::::::::::::::::::::

::::::::::::::::::::::::::::::::::::: objectives

- Explain

::::::::::::::::::::::::::::::::::::::::::::::::

## Variant Quality Score Recalibration

The raw VCF file from the previous step (`output.vcf.gz`) contains 10467 variants. Not all of these are real, therefore, the aim of this step is to filter out artifacts or false positive variants. GATK has provided different workflows for variant filtering. Here we will walk through the Variant Quality Score Recalibration or the VQSR strategy. VQSR is a two step process (1) the first step builds a model that describes how variant metric or quality measures co-vary with the known variants in the training set. (2) The second step then ranks each variant according to the target sensitivity cutoff and applies a filter expression.

::::::::::::::::::::::::::::::::::::: challenge 

#### Challenge 4.1

Step 1. VariantRecalibrator

```bash
gatk --java-options "-Xmx7g" VariantRecalibrator \
    -V output/output.vcf.gz \
    --trust-all-polymorphic \
    -mode SNP \
    --max-gaussians 6 \
    --resource:hapmap,known=false,training=true,truth=true,prior=15 reference/hg38/hapmap_3.3.hg38.vcf.gz \
    --resource:omni,known=false,training=true,truth=true,prior=12 reference/hg38/1000G_omni2.5.hg38.vcf.gz \
    --resource:1000G,known=false,training=true,truth=false,prior=10 reference/hg38/1000G_phase1.snps.high_confidence.hg38.vcf.gz \
    --resource:dbsnp,known=true,training=false,truth=false,prior=7 reference/hg38/dbsnp_138.hg38.vcf.gz \
    -an QD -an MQRankSum -an ReadPosRankSum -an FS -an MQ -an SOR -an DP \
    -O output/cohort_snps.recal \
    --tranches-file output/cohort_snps.tranches
```

Step 2. ApplyVQSR

```bash
gatk --java-options "-Xmx7g" ApplyVQSR \
    -R reference/hg38/Homo_sapiens_assembly38.fasta \
    -V output/output.vcf.gz \
    -O output/output.vqsr.vcf \
    --truth-sensitivity-filter-level 99.0 \
    --tranches-file output/cohort_snps.tranches \
    --recal-file output/cohort_snps.recal \
    -mode SNP
```

:::::::::::::::::::::::::::::::::

::::::::::::::::::::::::::::::::::::::: callout

#### Countvariants
There are number of ways to count the variants in a VCF file. A very straight forward way using the GATK4 tools is as follows:

```bash
gatk CountVariants -V output/output.vqsr.vcf
```

```output
Tool returned:
10467
```
::::::::::::::::::::::::::::::::::::::::::::::

There are several protocols for filtering VCF files. We have walked through just one of them. For other options please visit this [link](https://gatk.broadinstitute.org/hc/en-us/articles/360035531112--How-to-Filter-variants-either-with-VQSR-or-by-hard-filtering).

::::::::::::::::::::::::::::::::::::: challenge 

#### Filtering strategy for a single sample VCF file

Consider the following method to filter a single sample VCF file. Here we will go through the Convolutional Neural Net based protocol to annotate and filter the VCF file.

This is a two step process:

(i) CNNScoreVariants will annotate the variant with pre-computed single-sample derived model scores in the INFO field CNN_1D (the neural network performs convolutions over the reference sequence surrounding the variant and combines those features with a multilayer perceptron on the variant annotations).

```bash
gatk --java-options "-Xmx7g" CNNScoreVariants  \
   -R reference/hg38/Homo_sapiens_assembly38.fasta \
   -V output/output.vcf.gz \
   -O output/output.cnns.vcf
```

(ii) FilterVariantTranches takes as input the percent sensitivities (0-100) to known sites to apply the filter. Variants with scores higher than for e.g. 99th percentile of variants in the resources pass through the filter and will have `PASS` in the filter. Others will have a filter values like ‘CNN_1D_INDEL_Tranche_99.40_100.00’ or ‘CNN_1D_SNP_Tranche_99.95_100.00’.

```bash
gatk --java-options "-Xmx7g" FilterVariantTranches \
    -V output/output.cnns.vcf \
    --resource reference/hg38/hapmap_3.3.hg38.vcf.gz \
    --resource reference/hg38/Mills_and_1000G_gold_standard.indels.hg38.vcf.gz \
    --info-key CNN_1D \
    --snp-tranche 99.95 \
    --indel-tranche 99.4 \
    -O output/output.cnns.cnnfilter.vcf
```
:::::::::::::::::::::::::::::::::

::::::::::::::::::::::::::::::::::::::: callout

#### Note
[BCFtools](https://samtools.github.io/bcftools/) is a useful tool to manipulate, filter and query VCF files. It can be combined with linux command line tools as well to summarise data. For example, the command below can used extract and print the ‘FILTER’ column from the VCF file.

```bash
bcftools query -f'%FILTER\n' output/output.vqsr.vcf
```
::::::::::::::::::::::::::::::::::::::::::::::

## Additional filtering

The VariantFiltration tools is designed for hard-filtering variant calls based on custom quality criteria such as sequencing depth, mapping quality etc. The two parameters are the filter-name and filter-expression. The parameter filter-name is the name of the filter to be used in the FILTER column if the expression in filter-expression is true. In the example below, if the sequencing depth at the variant site (VCF field DP) is less than 10, the FILTER field will be populated with the value ‘Low_depth10’. Users can add multiple filter expression/name combinations.

::::::::::::::::::::::::::::::::::::: challenge 

#### Challenge 4.2

```bash
gatk --java-options "-Xmx7g" VariantFiltration \
    -R reference/hg38/Homo_sapiens_assembly38.fasta \
    -V output/output.vqsr.vcf \
    -O output/output.vqsr.varfilter.vcf \
    --filter-name "Low_depth10" \
    --filter-expression "DP < 10"
```
:::::::::::::::::::::::: solution 

## Q: How many variants have a low sequencing depth (DP<10) in the file output.vqsr.varfilter.vcf?

```bash
bcftools query -f'%FILTER\n' output/output.vqsr.varfilter.vcf | sort | uniq -c
```

```output
  6 Low_depth10
  2 Low_depth10;VQSRTrancheSNP99.00to99.90
  9 Low_depth10;VQSRTrancheSNP99.90to100.00
  9064 PASS
  1278 VQSRTrancheSNP99.00to99.90
  108 VQSRTrancheSNP99.90to100.00
```

:::::::::::::::::::::::::::::::::
:::::::::::::::::::::::::::::::::

## Final analysis ready VCF file

Given we have a filter annotated VCF files (), we can now create an analysis ready VCF file.

::::::::::::::::::::::::::::::::::::: challenge 

#### Challenge 4.3

Create a VCF file called `output/output.vqsr.varfilter.pass.vcf.gz` that contains only PASS variants. The input VCF file is `output/output.vqsr.varfilter.vcf`.

:::::::::::::::::::::::: solution 

## Hint: Try using the Bcftools application.

Use the bcftools to filter PASS variants.

```bash
bcftools view -f 'PASS,.' -O vcf -o output/output.vqsr.varfilter.pass.vcf output/output.vqsr.varfilter.vcf
```

We have now created an analysis ready version of the VCF file. It is also a good practice to compress and index the file.

```output
bgzip -c output/output.vqsr.varfilter.pass.vcf > output/output.vqsr.varfilter.pass.vcf.gz
tabix -p vcf output/output.vqsr.varfilter.pass.vcf.gz
```
:::::::::::::::::::::::::::::::::
:::::::::::::::::::::::::::::::::

::::::::::::::::::::::::::::::::::::: keypoints 

- Use 

::::::::::::::::::::::::::::::::::::::::::::::::

